Fix hang on vthread unmount when -XX:+YieldPinnedVirtualThreads is enabled

fengxue-IS opened the pull request:
- Enter/exit critical transition when preparing for unmount/re-mount pinned vthread.
- Fix Object.wait/notify logic.
- Allow pinned vthread to yield in normal path.

#20709 passed with this change using `-Xint -XX:+YieldPinnedVirtualThreads`

babsingh commented:
@fengxue-IS code review is in the top commit of https://github.com/babsingh/openj9/commits/jep491-4

can you update the PR description to reflect the three commits?

fengxue-IS commented:
Updated PR description, squashed commits, applied review comments.
Local compile/testing passed on xlinux and amac.

@babsingh can you start the PR builds please

babsingh commented:
jenkins test sanity.functional,sanity.openjdk xmac,amac,aix,plinux jdk24

babsingh commented:
jenkins test sanity.functional amac,plinux jdk21

babsingh commented:
jenkins compile win jdk21,jdk24

babsingh commented:
@fengxue-IS There are compilation failures on JDK21. Code needs to be correctly wrapped with JDK24 ifdefs.

babsingh commented on runtime/vm/BytecodeInterpreter.hpp:
```diff
@@ -5763,6 +5771,13 @@ class INTERPRETER_CLASS
 		buildInternalNativeStackFrame(REGISTER_ARGS);
 		updateVMStruct(REGISTER_ARGS);
 
+		if (J9_ARE_ANY_BITS_SET(_vm->extendedRuntimeFlags3, J9_EXTENDED_RUNTIME3_YIELD_PINNED_CONTINUATION)
+		&& (_currentThread->ownedMonitorCount > 0)
+		&& !isFinished
+		) {
+			preparePinnedVirtualThreadForUnmount(_currentThread, NULL, false);
+		}
+
```
This is probably the culprit.

fengxue-IS commented:
sanity.openjdk xmac failure https://openj9-jenkins.osuosl.org/job/Test_openjdk24_j9_sanity.openjdk_x86-64_mac_Personal_testList_2/3/testReport/
failed in JITed code, doesn't look related to this 
sanity.functional failure tracked by https://github.com/eclipse-openj9/openj9/issues/21303.

I've fixed the JDK21 compile issue and verified locally.

babsingh commented:
jenkins test sanity.functional,sanity.openjdk xmac,amac,aix,plinux jdk24

babsingh commented:
jenkins test sanity.functional amac,plinux jdk21

babsingh commented:
jenkins compile win jdk21,jdk24

babsingh commented:
@fengxue-IS can you create a PR for the 0.51 release milestone which includes the commits from this PR and https://github.com/eclipse-openj9/openj9/pull/21273?

babsingh commented:
> JDK24 failure on amac [java/lang/invoke/LoopCombinatorLongSignatureTest.java.LoopCombinatorLongSignatureTest](https://openj9-jenkins.osuosl.org/job/Test_openjdk24_j9_sanity.openjdk_aarch64_mac_Personal_testList_0/11/testReport/java_lang_invoke_LoopCombinatorLongSignatureTest/java/LoopCombinatorLongSignatureTest/)

I don't believe this issue has been reported before. Can you verify whether the test fails with and without your changes to determine if the failure is due to this PR? We might have unintentionally broken `Thread.initialize`. If the issue is unrelated, please open a separate Github issue for this failure.

fengxue-IS commented:
Confirmed failure exist in head branch via grinder https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/48649/
Build: https://hyc-runtimes-jenkins.swg-devops.com/job/Build_JDK24_aarch64_mac_Personal/44/
```
16:36:39  =JAVA VERSION OUTPUT BEGIN=
16:36:43  openjdk version "24-internal" 2025-03-18
16:36:43  OpenJDK Runtime Environment (build 24-internal-adhoc.jenkins.BuildJDK24aarch64macPersonal)
16:36:43  Eclipse OpenJ9 VM (build master-4f09a6d2e61, JRE 24 Mac OS X aarch64-64-Bit 20250307_44 (JIT enabled, AOT enabled)
16:36:43  OpenJ9   - 4f09a6d2e61
16:36:43  OMR      - ffd3ef9046b
16:36:43  JCL      - c8ca8c74788 based on jdk-24+36)
```
Will open a issue to document the failure.
